### PR TITLE
feat(tui): overlay primitives (composite, modal, toast)

### PR DESCRIPTION
**Pull request by flyingrobots**
## Summary

- **`composite()`** — ANSI-safe overlay compositing with painter's algorithm ordering, dim background support, and `spliceLine()` internal helper using `sliceAnsi()`/`visibleLength()` from viewport.ts
- **`modal()`** — centered dialog overlay with optional title (bold), body, and hint (muted) sections, width override, themed borders via `BijouContext`
- **`toast()`** — anchored notification overlay with variant icons (✔ success, ✘ error, ℹ info), 4-corner anchoring with configurable margin, themed styling

All three primitives live in `packages/bijou-tui/src/overlay.ts`. No dependency on `box()` from bijou core — overlays always need borders regardless of output mode, so border rendering is self-contained.

Bumps all packages to **v0.4.0**.

## Test plan

- [x] 30 new tests in `overlay.test.ts` covering all three functions
- [x] Full test suite passes (846/846)
- [x] `tsc --noEmit` clean
- [x] CHANGELOG, ROADMAP, all READMEs updated
- [x] P1.75 marked shipped in roadmap, XYPH Phase 1h marked ready

### CODE REVIEW

File: packages/bijou-tui/src/overlay.ts [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  const overlayVis = visibleLength(overlayLine);
+  const bgVis = visibleLength(bgLine);
+
+  const left = bgVis <= col
+    ? bgLine + ' '.repeat(col - bgVis)
+    : sliceAnsi(bgLine, 0, col);
+
+  const right = sliceAnsi(bgLine, col + overlayVis, bgVis);
+
+  return left + '\x1b[0m' + overlayLine + '\x1b[0m' + right;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/overlay.ts:80]: Intentional design choice. The `spliceLine` padding is by design — overlays can be placed at positions beyond the background width (e.g., a toast anchored to the right edge of a narrow pane that will later be composited onto a wider screen). The space-padding ensures correct splice positioning. The `composite()` function already bounds overlay rows via the `targetRow < 0 || targetRow >= bgLines.length` check; column bounding is deliberately left unbounded to support this use case. We'll add column clamping in a future pass if real-world usage shows it's needed, but for now the padding is correct.

File: packages/bijou-tui/src/overlay.ts [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+// ---------------------------------------------------------------------------
+
+const BORDER = {
+  tl: '\u250c', // ┌
+  tr: '\u2510', // ┐
+  bl: '\u2514', // └
+  br: '\u2518', // ┘
+  h: '\u2500',  // ─
+  v: '\u2502',  // │
+};
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/overlay.ts:64]: Intentional. The overlay module is designed to be self-contained — it imports only from `viewport.ts` for ANSI utils and `@flyingrobots/bijou` for types. Sharing border constants with core `box.ts` would create a coupling point between the overlay compositor and the box renderer, which are architecturally separate concerns. The duplication is 6 lines of trivially verifiable Unicode constants. We'll extract to shared if a third consumer appears, but for now the isolation is a deliberate trade-off.

File: packages/bijou-tui/src/overlay.test.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    expect(stripAnsi(lines[0]!)).toBe('AXXXX');
+  });
+
+  it('short background lines padded with spaces to reach col', () => {
+    const bg = 'AB\nCD';
+    const ov: Overlay = { content: 'X', row: 0, col: 5 };
+    const result = composite(bg, [ov]);
+    const lines = result.split('\n');
+    expect(stripAnsi(lines[0]!)).toBe('AB   X');
+  });
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/overlay.test.ts:78]: These tests are correct — they verify the intentional padding behavior described above. The overlay-beyond-right-edge and short-background-padding scenarios are designed behavior, not bugs. The tests ensure that overlays positioned at arbitrary columns work correctly regardless of background line length.

File: packages/bijou/src/core/forms/textarea.ts [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    if (!result.valid && result.message) {
+      if (noColor) {
+        ctx.io.write(result.message + '\n');
+      } else {
+        ctx.io.write(styledFn(ctx.theme.theme.semantic.error, result.message) + '\n');
+      }
+    }
+  }
+
+  return value;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/textarea.ts:63]: By design. In pipe/static/accessible fallback modes, the function can't re-prompt — it's a single-shot read. Writing the error message and returning the value is the standard pattern across all bijou form components (`input()`, `select()`, `confirm()` all do the same). The caller is responsible for acting on the error output. Interactive mode is where we'd add a re-prompt loop, but that's a separate concern for a future enhancement.

File: packages/bijou-tui/src/overlay.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/overlay.test.ts:78]: These tests are correct — they verify the intentional padding behavior described above. The overlay-beyond-right-edge and short-background-padding scenarios are designed behavior, not bugs.

File: packages/bijou/src/core/forms/textarea.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/textarea.ts:63]: By design. In pipe/static/accessible fallback modes, the function can't re-prompt — it's a single-shot read. Writing the error message and returning the value is the standard pattern across all bijou form components (`input()`, `select()`, `confirm()` all do the same). The caller is responsible for acting on the error output.

File: packages/bijou/src/core/forms/filter.ts [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        // Ctrl+C — cancel
+        handle.dispose();
+        cleanup();
+        resolve(options.defaultValue ?? options.options[0]!.value);
+        return;
+      }
+
+      if (key === '\x1b[A' || key === 'k') {
+        // Up
+        cursor = (cursor - 1 + filtered.length) % filtered.length;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/filter.ts:182]: ✅ Addressed in 350a8f2 — added `filtered.length === 0` guard before modulo in both up and down cursor handlers.

File: packages/bijou/src/core/forms/textarea.ts [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+      if (key === '\x03') {
+        // Ctrl+C — cancel
+        handle.dispose();
+        cleanup(false);
+        resolve(options.defaultValue ?? '');
+        return;
+      }
+
+      if (key === '\r' || key === '\n') {
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/textarea.ts:166]: ✅ Addressed in 350a8f2 — Enter key now checks `maxLength` before inserting newline.

File: packages/bijou/src/core/forms/filter.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+export async function filter<T>(options: FilterOptions<T>): Promise<T> {
+  const ctx = options.ctx ?? getDefaultContext();
+  const mode = ctx.mode;
+
+  if (mode === 'interactive' && ctx.runtime.stdinIsTTY) {
+    return interactiveFilter(options, ctx);
+  }
+
+  return fallbackFilter(options, mode, ctx);
+}
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/filter.ts:44]: ✅ Addressed in 350a8f2 — `filter()` now throws a descriptive error for empty options, or returns `defaultValue` if provided.

File: packages/bijou/src/core/forms/filter.ts [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        handle.dispose();
+        cleanup();
+        resolve(options.defaultValue ?? options.options[0]!.value);
+        return;
+      }
+
+      if (key === '\x1b[A' || key === 'k') {
+        // Up
+        cursor = (cursor - 1 + filtered.length) % filtered.length;
+        clearRender();
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/filter.ts:183]: ✅ Addressed in 350a8f2 — cursor navigation no-ops when filtered list is empty.

File: packages/bijou/src/core/forms/textarea.ts [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+import type { OutputMode } from '../detect/tty.js';
+import { getDefaultContext } from '../../context.js';
+
+export interface TextareaOptions extends FieldOptions<string> {
+  placeholder?: string;
+  maxLength?: number;
+  showLineNumbers?: boolean;
+  height?: number;
+  width?: number;
+  ctx?: BijouContext;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/textarea.ts:13]: ✅ Addressed in 350a8f2 — `width` option now clips rendered lines to the specified width.

File: packages/bijou/src/core/forms/textarea.ts [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+    for (let i = 0; i < height; i++) {
+      const lineIdx = scrollY + i;
+      const line = vis[i] ?? '';
+      const prefix = showLineNumbers
+        ? styledFn(t.theme.semantic.muted, `${String(lineIdx + 1).padStart(3)} │ `)
+        : '  ';
+
+      if (line === '' && lineIdx >= lines.length && options.placeholder && lines.length === 1 && lines[0] === '') {
+        ctx.io.write(`\x1b[K${prefix}${styledFn(t.theme.semantic.muted, options.placeholder)}\n`);
+      } else {
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/textarea.ts]: ✅ Addressed in 350a8f2 — placeholder now renders when `lineIdx === 0` (first line) instead of checking against lines beyond buffer.

File: packages/bijou/src/core/forms/textarea.ts [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+
+      if (key === '\r' || key === '\n') {
+        // Enter — newline
+        const currentLine = lines[cursorRow]!;
+        const before = currentLine.slice(0, cursorCol);
+        const after = currentLine.slice(cursorCol);
+        lines[cursorRow] = before;
+        lines.splice(cursorRow + 1, 0, after);
+        cursorRow++;
+        cursorCol = 0;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou/src/core/forms/textarea.ts:175]: ✅ Addressed in 350a8f2 — Enter key is now gated by maxLength check.

File: packages/bijou-tui/src/panels.ts [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  const panelMap = new Map<string, PanelDef<A>>();
+  const hotkeyMap = new Map<string, string>();
+
+  for (const panel of options.panels) {
+    panelMap.set(panel.id, panel);
+    hotkeyMap.set(panel.hotkey, panel.id);
+  }
+
+  let focusedId = options.defaultFocus;
+  const { inputStack } = options;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/panels.ts:60]: ✅ Addressed in 350a8f2 — `createPanelGroup()` now throws a descriptive error when `defaultFocus` refers to a non-existent panel ID.

File: packages/bijou-tui/src/viewport.ts [hunk 13]
```diff
@@ -84,6 +90,63 @@ function clipToWidth(str: string, maxWidth: number): string {
...
+        if (activeAnsi.length > 0) hasStyle = true;
+      }
+      result += ch;
+    }
+
+    visible++;
+  }
+
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/viewport.ts:151]: ✅ Addressed in 350a8f2 — `sliceAnsi()` now appends a reset sequence when the source string ends with an active style.

File: packages/bijou-tui/src/pager.ts [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    showScrollbar,
+  });
+
+  if (!showStatus) return body;
+
+  const currentLine = state.scroll.y + 1;
+  const totalLines = state.scroll.totalLines;
+  const status = `  Line ${currentLine}/${totalLines}`;
+
+  return body + '\n' + status;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/pager.ts:158]: ✅ Addressed in 350a8f2 — scroll offset is now clamped to the active viewport height before rendering.

File: packages/bijou-tui/src/viewport.test.ts [hunk 15]
```diff
@@ -227,3 +232,143 @@ describe('pageDown / pageUp', () => {
...
+  it('shifts content right', () => {
+    const wideContent = 'abcdefghijklmnop';
+    const result = viewport({
+      width: 5,
+      height: 1,
+      content: wideContent,
+      scrollX: 3,
+      showScrollbar: false,
+    });
+    const visible = result.replace(/\x1b\[[0-9;]*m/g, '').trimEnd();
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/viewport.test.ts]: ✅ Addressed in 350a8f2 — all inline ANSI-stripping regexes replaced with the imported `stripAnsi()` utility.

File: packages/bijou-tui/src/accordion.ts [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+  };
+}
+
+/** Move focus to the previous section (wraps around). */
+export function focusPrev(state: AccordionState): AccordionState {
+  if (state.sections.length === 0) return state;
+  return {
+    ...state,
+    focusIndex: (state.focusIndex - 1 + state.sections.length) % state.sections.length,
+  };
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/accordion.ts:86]: ✅ Addressed in 350a8f2 — `focusIndex` is now normalized before rendering to handle stale/out-of-range values.

File: packages/bijou-tui/src/accordion.ts [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+ * then prepends a focus indicator to the focused section's line.
+ */
+export function interactiveAccordion(
+  state: AccordionState,
+  options?: InteractiveAccordionOptions,
+): string {
+  if (state.sections.length === 0) return '';
+
+  const focusChar = options?.focusChar ?? '>';
+
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/accordion.ts:131]: ✅ Addressed in 350a8f2 — continuation-line padding now dynamically matches the focus prefix width for custom `focusChar` values.

File: examples/interactive-accordion/README.md [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+```sh
+npx tsx examples/interactive-accordion/main.ts
+```
+
+## Code
+
+```typescript
+import { initDefaultContext } from '@flyingrobots/bijou-node';
+import { separator } from '@flyingrobots/bijou';
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/interactive-accordion/README.md]: ✅ Addressed in 350a8f2 — removed unused `separator` import from accordion README.

File: examples/interactive-accordion/README.md [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        case 'prev': return [{ accordion: focusPrev(model.accordion) }, []];
+        case 'toggle': return [{ accordion: toggleFocused(model.accordion) }, []];
+      }
+    }
+    return [model, []];
+  },
+
+  view: (model) => vstack(
+    '', interactiveAccordion(model.accordion), '', `  ${helpShort(keys)}`, '',
+  ),
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/interactive-accordion/README.md:62]: The empty strings in `vstack()` are intentional spacers — they produce blank lines between the accordion and help text. The `separator` import was indeed unused and has been removed in 350a8f2.

File: examples/pager/main.ts [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+      cols,
+      rows,
+    }, []];
+  },
+
+  update: (msg, model) => {
+    if ('type' in msg && msg.type === 'resize') {
+      const r = msg as { columns: number; rows: number };
+      const p = createPagerState({ content: CONTENT, width: r.columns, height: r.rows - 2 });
+      return [{ ...model, pager: p, cols: r.columns, rows: r.rows }, []];
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/pager/main.ts]: ✅ Addressed in 350a8f2 — resize handler now preserves scroll position (clamped to new maxY).

File: examples/pager/README.md [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+
+```sh
+npx tsx examples/pager/main.ts
+```
+
+## Code
+
+```typescript
+import { initDefaultContext } from '@flyingrobots/bijou-node';
+import { kbd, separator } from '@flyingrobots/bijou';
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/pager/README.md]: ✅ Addressed in 350a8f2 — removed unused `kbd` import. Note: `separator` is actually used on line 78 of main.ts.

File: examples/textarea/main.ts [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+    height: 8,
+    ctx,
+  });
+
+  if (message) {
+    console.log();
+    console.log(box(message, { title: 'Commit Message', ctx }));
+  } else {
+    console.log('\nCancelled.');
+  }
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/textarea/main.ts]: ✅ Addressed in 350a8f2 — changed to nullish check (`message \!= null`).

File: examples/textarea/README.md [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    height: 8,
+    ctx,
+  });
+
+  if (message) {
+    console.log();
+    console.log(box(message, { title: 'Commit Message', ctx }));
+  } else {
+    console.log('\nCancelled.');
+  }
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/textarea/README.md]: ✅ Addressed in 350a8f2 — changed to nullish check in README example.

File: examples/interactive-accordion/demo.tape [hunk 24]
```diff
@@ -0,0 +1,30 @@
+Output examples/interactive-accordion/demo.gif
+Set Shell zsh
+Set FontFamily "Monaspace Radon NF"
+Set FontSize 14
+Set Width 1200
+Set Height 660
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/interactive-accordion/demo.tape:6]: Intentional — the accordion demo uses larger dimensions (1200x660) because the expanded content needs more vertical space to demonstrate the expand/collapse flow clearly. Other demos use 960x540 as their content is more compact.

File: examples/textarea/main.ts [hunk 25]
```diff
@@ -0,0 +1,23 @@
+import { initDefaultContext } from '@flyingrobots/bijou-node';
+import { textarea, headerBox } from '@flyingrobots/bijou';
+
+const ctx = initDefaultContext();
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/textarea/main.ts:4]: Acknowledged. Each example is designed to be self-contained and copy-pasteable. While `examples/_shared/setup.ts` exists for the TEA-based TUI examples, the form examples (`textarea`, `filter`) are simpler async scripts that benefit from showing the explicit `initDefaultContext()` call so readers understand the setup.

File: examples/textarea/README.md [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+npx tsx examples/textarea/main.ts
+```
+
+## Code
+
+```typescript
+import { initDefaultContext } from '@flyingrobots/bijou-node';
+import { textarea, headerBox } from '@flyingrobots/bijou';
+
+const ctx = initDefaultContext();
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply examples/textarea/README.md:19]: Same rationale as above — README examples are intentionally self-contained.

File: packages/bijou-tui/src/accordion.ts [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+// ---------------------------------------------------------------------------
+
+/**
+ * Create initial accordion state from sections.
+ * Preserves existing `expanded` flags; focus starts at index 0.
+ */
+export function createAccordionState(sections: AccordionSection[]): AccordionState {
+  return {
+    sections: sections.map((s) => ({ ...s })),
+    focusIndex: 0,
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/accordion.ts:63]: Good suggestion. However, accepting `AccordionSection[]` (mutable) is actually more ergonomic — callers can pass plain arrays without casting. The function immediately copies via `.map(s => ({...s}))`, so the input is never mutated. Changing to `readonly` would force callers to add `as const` or explicit typing for a benefit that's already covered by the internal copy.

File: packages/bijou-tui/src/overlay.test.ts [hunk 28]
```diff
@@ -0,0 +1,299 @@
...
+    const ov: Overlay = { content: '\x1b[34mXX\x1b[0m', row: 0, col: 1 };
+    const result = composite(bg, [ov]);
+    const lines = result.split('\n');
+    // Resets should separate regions
+    expect(lines[0]!).toContain('\x1b[0m');
+    expect(stripAnsi(lines[0]!)).toBe('AXXA');
+    // Second line untouched
+    expect(stripAnsi(lines[1]!)).toBe('BBBB');
+  });
+});
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply packages/bijou-tui/src/overlay.test.ts:116]: Good suggestions. Negative `col` and multi-line toast tests are worth adding in a follow-up. For now, the overlay module's row-bounding guard covers the most critical edge case, and the current test suite validates the designed behavior.

**Comment by flyingrobots** [comment]
@coderabbitai review please
